File: VindsidenApp/RHEStationPickerViewController.m
//
//  RHEStationPickerViewController.m
//  Vindsiden
//
//  Created by Ragnar Henriksen on 14.05.12.
//  Copyright (c) 2012 Shortcut AS. All rights reserved.
//

#import "RHCAppDelegate.h"
#import "RHEStationPickerViewController.h"
#import "RHEVindsidenAPIClient.h"
#import "CDStation.h"

@import VindsidenKit;

@interface RHEStationPickerViewController ()

@property (strong, nonatomic) NSFetchedResultsController *fetchedResultsController;
@property (assign, nonatomic) BOOL changeIsUserDriven;
@end

@implementation RHEStationPickerViewController
{
    BOOL _indexOffset;
}

@synthesize delegate = _delegate;
@synthesize fetchedResultsController = _fetchedResultsController;
@synthesize changeIsUserDriven = _changeIsUserDriven;


- (void)dealloc
{
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}


- (void)viewDidLoad
{
    [super viewDidLoad];

    self.navigationItem.rightBarButtonItem = self.editButtonItem;
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(preferredContentSizeChanged:) name:UIContentSizeCategoryDidChangeNotification object:nil];
}


- (void)viewDidUnload
{
    [super viewDidUnload];
}


- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];

    [[[RHEVindsidenAPIClient defaultManager] operationQueue] setSuspended:YES];
}


- (void)viewWillDisappear:(BOOL)animated
{
    [super viewWillDisappear:animated];
    [[[RHEVindsidenAPIClient defaultManager] operationQueue] setSuspended:NO];
}


- (BOOL)shouldAutorotateToInterfaceOrientation:(UIInterfaceOrientation)interfaceOrientation
{
    return (interfaceOrientation == UIInterfaceOrientationPortrait);
}


- (void)setEditing:(BOOL)editing animated:(BOOL)animated
{
    [super setEditing:editing animated:animated];
    
    self.navigationItem.leftBarButtonItem.enabled = !editing;
}


- (void)preferredContentSizeChanged:(NSNotification *)aNotification
{
    [self.tableView reloadData];
}


#pragma mark - Table view data source


- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return 2;
}


- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    @try {
        id <NSFetchedResultsSectionInfo> sectionInfo = [[self fetchedResultsController] sections][section];
        return [sectionInfo numberOfObjects];
    }
    @catch (NSException *exception) {
        return 0;
    }
}


- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    static NSString *CellIdentifier = @"StationCell";
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:CellIdentifier];

    [self configureCell:cell atIndexPath:indexPath];
    return cell;
}


- (void) configureCell:(UITableViewCell *)cell atIndexPath:(NSIndexPath *)indexPath
{
    CDStation *station = (CDStation *)[[self fetchedResultsController] objectAtIndexPath:indexPath];

    cell.textLabel.text = station.stationName;
    cell.detailTextLabel.text = station.city;
}


- (void)tableView:(UITableView *)tableView moveRowAtIndexPath:(NSIndexPath *)fromIndexPath toIndexPath:(NSIndexPath *)toIndexPath
{
    if ( fromIndexPath == toIndexPath ) {
        return;
    }

    _changeIsUserDriven = YES;

    NSManagedObjectContext *context = [self.fetchedResultsController managedObjectContext];

    NSMutableArray *arr = [NSMutableArray arrayWithArray:[[self.fetchedResultsController fetchedObjects] sortedArrayUsingComparator:^NSComparisonResult(CDStation *obj1, CDStation *obj2) {
        return [obj1.order compare:obj2.order];
    }]];
    CDStation *objToMove = [self.fetchedResultsController objectAtIndexPath:fromIndexPath];

    objToMove.isHidden = @(toIndexPath.section == 1);

    [arr removeObject:objToMove];
    [arr insertObject:objToMove atIndex:toIndexPath.row];

    int iVisible = 100;
    int iHidden = 200;
    for ( CDStation *station in arr ) {
        if ( [station.isHidden boolValue] ) {
            station.order = @(++iHidden);
        } else {
            station.order = @(++iVisible);
        }
    }

    NSError *error = NULL;
    if (![context save:&error]) {
        DLOG(@"%@", error);
    }
}


- (BOOL)tableView:(UITableView *)tableView canMoveRowAtIndexPath:(NSIndexPath *)indexPath
{
    return YES;
}


- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath
{
    return YES;
}


- (UITableViewCellEditingStyle)tableView:(UITableView *)tableView editingStyleForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return UITableViewCellEditingStyleNone;
}


- (NSIndexPath *)tableView:(UITableView *)tableView targetIndexPathForMoveFromRowAtIndexPath:(NSIndexPath *)sourceIndexPath toProposedIndexPath:(NSIndexPath *)proposedDestinationIndexPath
{
    if ( sourceIndexPath.section == 1 ) {
        return proposedDestinationIndexPath;
    }

    NSInteger rows = [self tableView:self.tableView numberOfRowsInSection:sourceIndexPath.section];
    if ( rows <= 1 ) {
        return sourceIndexPath;
    }
    return proposedDestinationIndexPath;
}


#pragma mark - Table view delegate


- (BOOL)tableView:(UITableView *)tableView shouldIndentWhileEditingRowAtIndexPath:(NSIndexPath *)indexPath
{
    return NO;
}


- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section
{
    if ( section == 0 ) {
        return NSLocalizedString(@"Visible", nil);
    }
    return NSLocalizedString(@"Not Visible", nil);
}


- (void)tableView:(UITableView *)tableView willDisplayCell:(UITableViewCell *)cell forRowAtIndexPath:(NSIndexPath *)indexPath
{
    cell.textLabel.font = [UIFont preferredFontForTextStyle:[[cell.textLabel.font fontDescriptor] objectForKey:@"NSCTFontUIUsageAttribute"]];
    cell.detailTextLabel.font = [UIFont preferredFontForTextStyle:[[cell.detailTextLabel.font fontDescriptor] objectForKey:@"NSCTFontUIUsageAttribute"]];
}


#pragma mark - Fetched results controller


- (void)controllerWillChangeContent:(NSFetchedResultsController *)controller
{
    if ( !_changeIsUserDriven ) {
        [self.tableView beginUpdates];
    }
}


- (void)controller:(NSFetchedResultsController *)controller didChangeSection:(id <NSFetchedResultsSectionInfo>)sectionInfo
           atIndex:(NSUInteger)sectionIndex forChangeType:(NSFetchedResultsChangeType)type
{

    if ( _changeIsUserDriven ) {
        return;
    }

    switch(type) {
        case NSFetchedResultsChangeInsert:
            [self.tableView insertSections:[NSIndexSet indexSetWithIndex:sectionIndex]
                          withRowAnimation:UITableViewRowAnimationFade];
            break;

        case NSFetchedResultsChangeDelete:
            [self.tableView deleteSections:[NSIndexSet indexSetWithIndex:sectionIndex]
                          withRowAnimation:UITableViewRowAnimationFade];
            break;
    }
}


- (void)controller:(NSFetchedResultsController *)controller didChangeObject:(id)anObject
       atIndexPath:(NSIndexPath *)indexPath forChangeType:(NSFetchedResultsChangeType)type
      newIndexPath:(NSIndexPath *)newIndexPath
{

    if ( _changeIsUserDriven ) {
        return;
    }

    UITableView *tableView = self.tableView;

    switch(type) {

        case NSFetchedResultsChangeInsert:
            [tableView insertRowsAtIndexPaths:@[newIndexPath]
                             withRowAnimation:UITableViewRowAnimationTop];
            break;

        case NSFetchedResultsChangeDelete:
            [tableView deleteRowsAtIndexPaths:@[indexPath]
                             withRowAnimation:UITableViewRowAnimationTop];
            break;

        case NSFetchedResultsChangeUpdate:
            [self configureCell:[tableView cellForRowAtIndexPath:indexPath]
                    atIndexPath:indexPath];
            break;

        case NSFetchedResultsChangeMove:
            [tableView deleteRowsAtIndexPaths:@[indexPath]
                             withRowAnimation:UITableViewRowAnimationFade];
            [tableView insertRowsAtIndexPaths:@[newIndexPath]
                             withRowAnimation:UITableViewRowAnimationFade];
            break;
    }
}


- (void)controllerDidChangeContent:(NSFetchedResultsController *)controller
{
    if ( !_changeIsUserDriven ) {
        [self.tableView endUpdates];
    } else if ( !self.tableView.editing) {
        [self.tableView reloadData];
    }
    _changeIsUserDriven = NO;
}


- (NSFetchedResultsController *) fetchedResultsController
{
    if ( _fetchedResultsController != nil ) {
        return _fetchedResultsController;
    }

    NSManagedObjectContext *context = [[Datamanager sharedManager] managedObjectContext];
    NSString *cacheName = @"StationPicker";
    [NSFetchedResultsController deleteCacheWithName:cacheName];
    NSFetchRequest *fetchRequest = [[NSFetchRequest alloc] init];
    NSEntityDescription *entity = [NSEntityDescription entityForName:@"CDStation" inManagedObjectContext:context];
    [fetchRequest setEntity:entity];
    [fetchRequest setFetchBatchSize:20];

    NSSortDescriptor *sortDescriptor = [[NSSortDescriptor alloc] initWithKey:@"order" ascending:YES];
    NSArray *sortDescriptors = @[sortDescriptor];
    [fetchRequest setSortDescriptors:sortDescriptors];

    NSFetchedResultsController *aFetchedResultsController = [[NSFetchedResultsController alloc] initWithFetchRequest:fetchRequest
                                                                                                managedObjectContext:context
                                                                                                  sectionNameKeyPath:@"isHidden"
                                                                                                           cacheName:cacheName];
    aFetchedResultsController.delegate = self;
    self.fetchedResultsController = aFetchedResultsController;

    NSError *error = nil;
    if ( ! [_fetchedResultsController performFetch:&error]) {
        WARNING(@"Unresolved error %@, %@", error, [error userInfo]);
        abort();
    }

    return _fetchedResultsController;
}


#pragma mark - Actions


- (IBAction)done:(id)sender
{
   // [self dismissViewControllerAnimated:YES completion:nil];
    [self.delegate rheStationPickerDidFinish:self];
}


@end

File: VindsidenApp/RHEStationPickerViewController.swift
//
//  RHEStationPickerViewController.swift
//  VindsidenApp
//
//  Created by Ragnar Henriksen on 05/09/14.
//  Copyright (c) 2014 RHC. All rights reserved.
//

import UIKit
import CoreData
import VindsidenKit



class RHEStationPickerViewController : UITableViewController, NSFetchedResultsControllerDelegate
{
    var changeIsUserDriven = false


    deinit {
        NotificationCenter.default.removeObserver(self)
    }


    override func viewDidLoad() {
        super.viewDidLoad()

        navigationItem.rightBarButtonItem = self.editButtonItem
        NotificationCenter.default.addObserver(self, selector: #selector(RHEStationPickerViewController.preferredContentSizeDidChange(_:)), name: NSNotification.Name.UIContentSizeCategoryDidChange, object: nil)
    }


    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
    }


    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
    }


    override func setEditing(_ editing: Bool, animated: Bool) {
        super.setEditing(editing, animated: animated)
        navigationItem.leftBarButtonItem?.isEnabled = !editing
    }


    @objc func preferredContentSizeDidChange( _ notification: Notification ) {
        tableView.reloadData()
    }


    func configureCell( _ cell: UITableViewCell, atIndexPath indexPath: IndexPath) {
        let station = fetchedResultsController.object(at: indexPath) as! CDStation
        cell.textLabel?.text = station.stationName;
        cell.detailTextLabel?.text = station.city;

        if let hidden = station.isHidden, hidden == true {
            cell.imageView?.image = UIImage(named: "uncheckmark_icon", in: nil, compatibleWith: self.traitCollection)
        } else {
            cell.imageView?.image = UIImage(named: "checkmark_icon", in: nil, compatibleWith: self.traitCollection)
        }
    }


    // MARK: UITableView


    override func numberOfSections(in tableView: UITableView) -> Int {
        if let sections = fetchedResultsController.sections {
            return sections.count
        }

        return 0
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        if let sections = fetchedResultsController.sections {
            let sectionInfo = sections[section] as NSFetchedResultsSectionInfo

            return sectionInfo.numberOfObjects
        }

        return 0
    }


    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "StationCell", for: indexPath) as UITableViewCell

        configureCell(cell, atIndexPath: indexPath)

        return cell;
    }


    override func tableView(_ tableView: UITableView, canEditRowAt indexPath: IndexPath) -> Bool {
        return true
    }


    override func tableView(_ tableView: UITableView, canMoveRowAt indexPath: IndexPath) -> Bool {
        return true
    }


    override func tableView(_ tableView: UITableView, editingStyleForRowAt indexPath: IndexPath) -> UITableViewCellEditingStyle {
        return .none
    }


    override func tableView(_ tableView: UITableView, shouldIndentWhileEditingRowAt indexPath: IndexPath) -> Bool {
        return false
    }


    override func tableView(_ tableView: UITableView, moveRowAt sourceIndexPath: IndexPath, to destinationIndexPath: IndexPath) {
        if sourceIndexPath == destinationIndexPath {
            return
        }

        changeIsUserDriven = true

        var array = fetchedResultsController.fetchedObjects as! [CDStation]
        let objectToMove = fetchedResultsController.object(at: sourceIndexPath) as! CDStation
        array.remove(at: sourceIndexPath.row)
        array.insert(objectToMove, at: destinationIndexPath.row)

        var index = 0
        for object in array {
            object.order = NSNumber(value: index)
            index += 1
        }

        let context = DataManager.shared.viewContext()

        do {
            try context.save()
        } catch let error as NSError {
            print("Save failed: \(error.localizedDescription)")
        }
    }


    override func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath) {
        cell.textLabel?.font = UIFont.preferredFont(forTextStyle: (UIFontTextStyle(rawValue: cell.textLabel?.font.fontDescriptor.object(forKey: UIFontDescriptor.AttributeName(rawValue: "NSCTFontUIUsageAttribute")) as! String)))
        cell.detailTextLabel?.font = UIFont.preferredFont(forTextStyle: (UIFontTextStyle(rawValue: cell.detailTextLabel?.font.fontDescriptor.object(forKey: UIFontDescriptor.AttributeName(rawValue: "NSCTFontUIUsageAttribute")) as! String)))
    }


    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let station = fetchedResultsController.object(at: indexPath) as! CDStation

        if let hidden = station.isHidden {
            station.isHidden = NSNumber(value: !hidden.boolValue as Bool)
        }

        if let hidden = station.isHidden, hidden.boolValue == false {
            DataManager.shared.addStationToIndex(station)
        } else {
            DataManager.shared.removeStationFromIndex(station)
        }


        let context = DataManager.shared.viewContext()

        do {
            try context.save()
        } catch let error as NSError {
            print("Save failed: \(error.localizedDescription)")
        }

        tableView.deselectRow(at: indexPath, animated: true)
    }


    // MARK: NSFetchedResultsController


    lazy var fetchedResultsController : NSFetchedResultsController<NSFetchRequestResult> = {
        NSFetchedResultsController<NSFetchRequestResult>.deleteCache(withName: "StationPicker")

        let contxt = DataManager.shared.viewContext()

        let fetchRequest = CDStation.fetchRequest()
        fetchRequest.fetchBatchSize = 20
        fetchRequest.sortDescriptors = [NSSortDescriptor(key: "order", ascending: true)]

        let controller = NSFetchedResultsController(fetchRequest: fetchRequest, managedObjectContext: contxt, sectionNameKeyPath: nil, cacheName: "StationPicker")
        controller.delegate = self

        do {
            try controller.performFetch()
        } catch {
            NSLog("Fetching stations failed")
            abort()
        }

        return controller
    }()


    func controllerWillChangeContent(_ controller: NSFetchedResultsController<NSFetchRequestResult>) {
        if !changeIsUserDriven {
            tableView.beginUpdates()
        }
    }

    func controller(_ controller: NSFetchedResultsController<NSFetchRequestResult>, didChange anObject: Any, at indexPath: IndexPath?, for type: NSFetchedResultsChangeType, newIndexPath: IndexPath?) {
        if changeIsUserDriven {
            return
        }

        switch (type) {
        case .insert:
            tableView.insertRows(at: [newIndexPath!], with: .top)
        case .delete:
            tableView.deleteRows(at: [indexPath!], with: .top)
        case .update:
            if let cell = tableView.cellForRow(at: indexPath!) {
                configureCell(cell, atIndexPath: indexPath!)
            }
        case .move:
            tableView.deleteRows(at: [indexPath!], with: .fade)
            tableView.insertRows(at: [newIndexPath!], with: .fade)
        }
    }


    func controllerDidChangeContent(_ controller: NSFetchedResultsController<NSFetchRequestResult>) {
        if !changeIsUserDriven {
            tableView.endUpdates()
        } else {
            tableView.reloadData()
        }

        changeIsUserDriven = false
    }
}

